### PR TITLE
Add new async tests, fix bug

### PR DESCRIPTION
**Pull request by reginleif888**

### CODE REVIEW

**Comment by reginleif888** [comment]
The test output before the fix:

```
Running test/handler/async.test.mo 
 ✖ 
 FAIL test/handler/async.test.mo:175:1: execution error, assertion failure
   ...
   174  | assert_state(35, 5, 1); // consolidation failed with updated deposit scheduled
-> 175  | assert handler.journalLength() == inc(4); // #consolidationError, #debited, #feeUpdated, #credited
   176  | assert (await* handler.notify(user1)) == ?(0, 14); // credit corrected after consolidation
```
   
The problem was that when the fee was increased while deposit is being consolidated (old_fee < deposit < new_fee), `queuedFunds` was not updated after `credit()`. So the balances integrity was broken (`#error` logged).

Fix:

```diff          
credit(p, deposit - expected_fee);
+ queuedFunds += deposit;
```

File: test/handler/async.test.mo [hunk 1]
```diff
@@ -71,85 +16,163 @@ func assert_state(x : (Nat, Nat, Nat)) {
...
+assert handler.journalLength() == inc(3); // #debited (recalculation), #newDeposit, #credited
+print("tree lookups = " # debug_show handler.lookups());
+
+// call multiple notify() simultaneously
+await ledger.lock_balance();
+let arr = [async { await* handler.notify(user1) }, async { await* handler.notify(user1) }, async { await* handler.notify(user1) }];
+await ledger.release_balance(); // let notify return
+assert (await arr[0]) == ?(0, 5); // first notify() should return state
+assert (await arr[1]) == null; // should return null
+assert (await arr[2]) == null; // should return null
```
Comment (timohanke) [inline test/handler/async.test.mo:128]: I would do this order:
```
assert (await arr[1]) == null; // should return null
assert (await arr[2]) == null; // should return null
await ledger.release_balance(); // let notify return
assert (await arr[0]) == ?(0, 5); // first notify() should return state
```
The null responses come immediately, they don't need the release_balance() call.

File: test/handler/async.test.mo [hunk 2]
```diff
@@ -71,85 +16,163 @@ func assert_state(x : (Nat, Nat, Nat)) {
...
+let f4 = async { await* handler.trigger() };
+await f3;
+await f4;
+assert ((await ledger.transfer_count())) == transfer_count + 1; // only 1 transfer call has been made
+assert_state(0, 5, 0); // consolidation successful
+assert handler.journalLength() == inc(1); // #consolidated
+print("tree lookups = " # debug_show handler.lookups());
+
+// increase fee while deposit is being consolidated
+// scenario 1: old_fee < deposit < new_fee
```
Comment (timohanke) [inline test/handler/async.test.mo:145]: `old_fee < deposit <= new_fee` or ?
Should also work with set_fee(20)

File: test/handler/async.test.mo [hunk 3]
```diff
@@ -71,85 +16,163 @@ func assert_state(x : (Nat, Nat, Nat)) {
...
+assert (await* handler.notify(user1)) == ?(20, 10); // deposit = 20, credit = 10
+assert handler.journalLength() == inc(2); // #credited, #newDeposit
+assert_state(20, 5, 1);
+await ledger.lock_transfer();
+let f5 = async { await* handler.trigger() };
+await ledger.set_fee(25);
+await ledger.set_response(#Err(#BadFee { expected_fee = 25 }));
+await ledger.release_transfer(); // let transfer return
+await f5;
+assert_state(0, 5, 0); // consolidation failed
```
Comment (timohanke) [inline test/handler/async.test.mo:157]: We should make one more assertion here, that the user's credit has been reduced (debited).

File: test/handler/async.test.mo [hunk 4]
```diff
@@ -71,85 +16,163 @@ func assert_state(x : (Nat, Nat, Nat)) {
...
+assert (await* handler.notify(user1)) == ?(35, 15); // deposit = 35, credit = 15
+assert handler.journalLength() == inc(2); // #credited, #newDeposit
+assert_state(35, 5, 1);
+await ledger.lock_transfer();
+let f6 = async { await* handler.trigger() };
+await ledger.set_fee(26);
+await ledger.set_response(#Err(#BadFee { expected_fee = 26 }));
+await ledger.release_transfer(); // let transfer return
+await f6;
+assert_state(35, 5, 1); // consolidation failed with updated deposit scheduled
```
Comment (timohanke) [inline test/handler/async.test.mo:174]: We should make one more assertion here, that the user's credit has been reduced (debited).

**Review by timohanke — COMMENTED** [review]
The really interesting test is when the fee change is _detected_ by the handler while a consolidation is underway. Currently, we only test when the fee change _happens_ while the consolidation is underway. The handler only detects the fee change when the consolidation returns (with BadError). So it gets detected at the same time when the consolidation returns. That's the less risky case.

The more risky case is when the the handler detects the fee change _before_ the consolidation returns. We must call update_fee() before releasing the consolidation to achieve that scenario. These are additional tests we need on top of the ones with detection via BadFee.

File: src/TokenHandler/AccountManager.mo [hunk 5]
```diff
@@ -23,7 +23,10 @@ module {
     #feeUpdated : { old : Nat; new : Nat };
     #newDeposit : Nat;
     #consolidated : { deducted : Nat; credited : Nat };
-    #consolidationError : ICRC1.TransferError or { #CallIcrc1LedgerError };
+    #consolidationError : ICRC1.TransferError or {
+      #CallIcrc1LedgerError;
+      #ExtendedBadFee : { original_fee : Nat; expected_fee : Nat };
+    };
```
Comment (timohanke) [inline src/TokenHandler/AccountManager.mo:29]: This makes the code too complicated. Please don't report the original fee from the inner function `processConsolidationTransfer` to the outer function `consolidate`. Keep their interaction as before. The outer function can save original fee itself for use in the continuation. Just put a line 
```
let originalFee : Nat = fee_;
```
in `consolidate` before it calls `processConsolidationTransfer`.

File: test/handler/async.test.mo [hunk 6]
```diff
@@ -248,3 +248,24 @@ assert (await fn8) == #err(#TooLowQuantity);
...
+let f9 = async { await* handler.trigger(); await ledger.set_balance(0) };
+await ledger.set_fee(5);
+ignore await* handler.updateFee();
+assert handler.journalLength() == inc(1); // #feeUpdated
+await ledger.set_response([#Err(#BadFee { expected_fee = 5 })]);
+await ledger.release_transfer(); // let transfer return
+await f9;
+assert_state(10, 5, 1); // consolidation failed
+assert handler.journalLength() == inc(3); // #consolidationError, #debited, #credited
+assert (await* handler.notify(user1)) == ?(0, 10); // credit has been corrected with new fee
```
Comment (timohanke) [inline test/handler/async.test.mo]: Better to call `handler.info(user1)` here. We're are targeting credit here, right?

File: src/TokenHandler/AccountManager.mo [hunk 7]
```diff
@@ -160,16 +163,18 @@ module {
       #Ok : Nat;
       #Err : ICRC1.TransferError or {
         #CallIcrc1LedgerError;
+        #ExtendedBadFee : { original_fee : Nat; expected_fee : Nat };
       };
     } {
       let transferAmount : Nat = Int.abs(deposit - fee_);
+      let originalFee : Nat = fee_;
```
Comment (timohanke) [inline src/TokenHandler/AccountManager.mo:170]: We can move this line entirely to the `consolidate()` function.

File: src/TokenHandler/AccountManager.mo [hunk 8]
```diff
@@ -183,6 +188,14 @@ module {
           totalConsolidated_ += transferAmount;
           log(p, #consolidated({ deducted = deposit; credited = transferAmount }));
         };
+        case (#Err(#BadFee({ expected_fee }))) {
+          let err = #ExtendedBadFee({
+            original_fee = originalFee;
+            expected_fee = expected_fee;
+          });
+          log(p, #consolidationError(err));
+          return #Err(err);
```
Comment (timohanke) [inline src/TokenHandler/AccountManager.mo:197]: We don't need to log the #ExtendedBadFee anymore, so we can remove #ExtendedBadFee entirely. Logging the consolidation failure is enough. We are already logging old and new fee inside `setNewFee` with #feeUpdated.

**Pull request merged**